### DISCUSSION TITLE
Cypress flutter integration

### DISCUSSION THREAD

**Discussion by sumanth-varma-us** [cypress-io/cypress#28477] [Questions and Help]
I am currently leveraging Cypress for the automated testing of a Flutter web application. Unfortunately, I am facing difficulties as Cypress is unable to retrieve certain widgets during testing. Could you please provide guidance on the proper approach to integrate Flutter with Cypress and effectively test the application?

Thanks in advance.

**Comment by portyu9** [comment]
Hi @sumanth-varma-us 

Cypress can test a Flutter **web** application, but there is an important limitation: Cypress queries the browser DOM, while Flutter renders most of its UI onto a canvas. Flutter widgets are therefore not automatically normal DOM elements that Cypress can locate.

The supported way to expose meaningful elements to browser tools is Flutter's **Semantics/accessibility layer**.

For example, you can enable web semantics programmatically:

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/semantics.dart';

void main() {
  runApp(const MyApp());

  if (kIsWeb) {
    SemanticsBinding.instance.ensureSemantics();
  }
}
```

Also make sure important controls expose useful semantics. Standard Flutter controls often do this already, while custom widgets may need `Semantics` labels/roles.

Then Cypress can interact with the HTML accessibility representation using stable roles/labels rather than trying to find Flutter's internal widget hierarchy.

For example, depending on the DOM Flutter generates:

```js
cy.get('[role="button"][aria-label="Submit"]').click()
```

For tests that need direct access to Flutter widgets, widget state, finders, or the Flutter rendering tree, Flutter's own `integration_test` tooling is the better fit. A good split is:

- **Flutter integration/widget testing** for Flutter-specific internals.
- **Cypress E2E testing** for the browser-visible workflow, URLs, network behavior, and accessible UI.

Trying to locate arbitrary canvas-rendered widgets with normal CSS selectors will not be reliable.

Please mark this answer as accepted if it helped, thank you.
